DOC: use substitution decorator for business month classes (#25828)

Batalex opened the pull request:
- [X] closes #25828 
- [ ] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry


Validation script
`CustomBusinessMonthEnd`
```
################################################################################
########## Docstring (pandas.tseries.offsets.CustomBusinessMonthEnd)  ##########
################################################################################
                                                                                
DateOffset subclass representing one custom business month, incrementing        
between end of month dates.                                                     
                                                                                
Parameters                                                                      
----------                                                                      
n : int, default 1                                                              
normalize : bool, default False                                                 
    Normalize start/end dates to midnight before generating date range          
weekmask : str, Default 'Mon Tue Wed Thu Fri'                                   
    weekmask of valid business days, passed to ``numpy.busdaycalendar``         
holidays : list                                                                 
    list/array of dates to exclude from the set of valid business days,         
    passed to ``numpy.busdaycalendar``                                          
calendar : pd.HolidayCalendar or np.busdaycalendar                              
offset : timedelta, default timedelta(0)                                        
                                                                                
################################################################################
################################## Validation ##################################
################################################################################
                                                                                
9 Errors found:                                                                 
        Summary should fit in a single line                                     
        Parameter "n" has no description                                        
        Parameter "normalize" description should finish with "."                
        Parameter "weekmask" description should start with a capital letter     
        Parameter "weekmask" description should finish with "."                 
        Parameter "holidays" description should start with a capital letter     
        Parameter "holidays" description should finish with "."                 
        Parameter "calendar" has no description                                 
        Parameter "offset" has no description                                   
2 Warnings found:                                                               
        See Also section not found                                              
        No examples section found                                                                                                                   
```

`CustomBusinessMonthBegin`
```
################################################################################
######### Docstring (pandas.tseries.offsets.CustomBusinessMonthBegin)  #########
################################################################################
                                                                                
DateOffset subclass representing one custom business month, incrementing        
between beginning of month dates.                                               
                                                                                
Parameters                                                                      
----------                                                                      
n : int, default 1                                                              
normalize : bool, default False                                                 
    Normalize start/end dates to midnight before generating date range          
weekmask : str, Default 'Mon Tue Wed Thu Fri'                                   
    weekmask of valid business days, passed to ``numpy.busdaycalendar``         
holidays : list                                                                 
    list/array of dates to exclude from the set of valid business days,         
    passed to ``numpy.busdaycalendar``                                          
calendar : pd.HolidayCalendar or np.busdaycalendar                              
offset : timedelta, default timedelta(0)                                        
                                                                                
################################################################################
################################## Validation ##################################
################################################################################
                                                                                
9 Errors found:                                                                 
        Summary should fit in a single line                                     
        Parameter "n" has no description                                        
        Parameter "normalize" description should finish with "."                
        Parameter "weekmask" description should start with a capital letter     
        Parameter "weekmask" description should finish with "."                 
        Parameter "holidays" description should start with a capital letter     
        Parameter "holidays" description should finish with "."                 
        Parameter "calendar" has no description                                 
        Parameter "offset" has no description                                   
2 Warnings found:                                                               
        See Also section not found                                              
        No examples section found                                               
```

Batalex committed:
DOC: use substitution decorator for business month classes (#25828)

WillAyd commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -35,6 +35,7 @@ Other Enhancements
 - :class:`RangeIndex` has gained :attr:`~RangeIndex.start`, :attr:`~RangeIndex.stop`, and :attr:`~RangeIndex.step` attributes (:issue:`25710`)
 - :class:`datetime.timezone` objects are now supported as arguments to timezone methods and constructors (:issue:`25065`)
 - :meth:`DataFrame.query` and :meth:`DataFrame.eval` now supports quoting column names with backticks to refer to names with spaces (:issue:`6508`)
+- :class:`CustomBusinessMonthBegin` and :class:`CustomBusinessMonthEnd` now use the substitution pattern for docstring sharing
```
This is an internal refactor and not visible to end users so no whatsnew required

Batalex replied:
Removed !

WillAyd reviewed (changes requested):
Thanks for the PR! Substitution looks correct. The docstring errors are orthogonal but do you see an easy way to fix them? May be worth knocking out as part of this if simple

Batalex commented:
@WillAyd I juste force pushed somes changes:
- Removed uneeded entry in whatsnew
- Fixed isort linting
- Fixed docstring validation. I picked some description elements from offsets.DateOffset & np.busdaycalendar

WillAyd reviewed (approved):
Nice job! FYI for next contribution shouldn't need to force push changes so long as you merge in master

WillAyd commented on pandas/tseries/offsets.py:
```diff
@@ -972,21 +972,25 @@ class BusinessMonthBegin(MonthOffset):
-    between [BEGIN/END] of month dates.
+    DateOffset subclass representing custom business month(s).
+
+    The count increase between %(bound)s of month dates.
```
Hmm actually can you retain the original wording here? Something like:

```suggestion
    Increments between %(bound)s of month dates.
```

Batalex replied:
Fixed!

WillAyd commented on pandas/tseries/offsets.py:
```diff
@@ -972,21 +972,25 @@ class BusinessMonthBegin(MonthOffset):
     calendar : pd.HolidayCalendar or np.busdaycalendar
+        A business day calendar.
     offset : timedelta, default timedelta(0)
+        Applies an offset counted in valid days.
```
Also I don't think an offset needs to be restricted to days - can you reword this? cc @jbrockmendel if you have any suggested wording for this parameter

Batalex replied:
Agreed. It seemed a bit strange to apply a millisecond offset to a month but still it is possible. I changed to a more neutral form if you do not have a better wording for now :)

Batalex committed:
Merge remote-tracking branch 'upstream/master' into doc/substitution-offset

WillAyd commented on pandas/tseries/offsets.py:
```diff
@@ -971,21 +971,25 @@ class BusinessMonthBegin(MonthOffset):
     calendar : pd.HolidayCalendar or np.busdaycalendar
+        A business day calendar.
     offset : timedelta, default timedelta(0)
+        Applies a time offset (from ms to years).
```
Still nitpicking but Python time delta only goes up to days so I think this is a misnomer saying years.

Probably better just to say something like `Time offset to apply` and leave any precision questions left to other parts of the documentation

Batalex replied:
Reworded.
I assumed that the offset could be in years because `pandas.Timedelta` can and is said to be interchangeable with it.
Better safe than sorry

WillAyd commented on pandas/tseries/offsets.py:
```diff
@@ -971,21 +971,25 @@ class BusinessMonthBegin(MonthOffset):
+        List/array of dates to exclude from the set of valid business days,
+        passed to ``numpy.busdaycalendar``.
     calendar : pd.HolidayCalendar or np.busdaycalendar
+        A business day calendar.
```
I suppose limiting this to "business day calendar" is also incorrect given the holiday calendar is a valid argument. Better to just leave generic and say something like `Calendar to integrate` - can always be made more fine grained in subsequent PRs just don't want it to be wrong

WillAyd reviewed (changes requested):
Very minor stuff. If you can clean up and ping on green can get this merged

WillAyd commented:
@Batalex just to clarify previous comment force pushing is less than desirable. Assuming you are rebasing which requires a force push an alternate workflow on top of your local branch is:

```sh
git fetch upstream
git merge upstream/master
```

Resolve any conflicts and `git merge --continue` if applicable. Should then be able to push without forcing

Batalex committed:
Merge remote-tracking branch 'upstream/master' into doc/substitution-offset

Batalex committed:
DOC: reword custom business month offset args descriptions

WillAyd reviewed (approved):
lgtm ping on green

Batalex commented:
@WillAyd ping

WillAyd commented:
Thanks @Batalex !

Batalex commented:
Thanks for your guidance!
